File: typescript/maroctax/src/simulate-raise.ts
import { calculateNetSalary } from './net-salary.js';
import type { FiscalYear, RaiseSimulation } from './types.js';
import { roundMoney } from './utils.js';

/**
 * Compares net salary before and after a gross raise.
 * @param currentGross - Current monthly gross in MAD.
 * @param newGross - Proposed monthly gross in MAD.
 * @param dependents - Number of dependents.
 * @param year - Fiscal year (defaults to 2025).
 */
export function simulateRaise(
  currentGross: number,
  newGross: number,
  dependents = 0,
  year: FiscalYear = 2025,
): RaiseSimulation {
  const before = calculateNetSalary(currentGross, dependents, year);
  const after = calculateNetSalary(newGross, dependents, year);
  const deltaGross = roundMoney(newGross - currentGross);
  const deltaNet = roundMoney(after.net - before.net);
  const deltaIr = roundMoney(after.ir - before.ir);
  const marginalNetRate = deltaGross === 0 ? 0 : roundMoney(deltaNet / deltaGross);

  return {
    year,
    dependents,
    before,
    after,
    deltaGross,
    deltaNet,
    deltaIr,
    marginalNetRate,
  };
}

File: dart/maroctax/lib/src/simulate_raise.dart
import 'net_salary.dart';
import 'types.dart';
import 'utils.dart';

/// Compares net salary before and after a gross raise.
RaiseSimulation simulateRaise(
  double currentGross,
  double newGross,
  int dependents, [
  FiscalYear year = FiscalYear.y2025,
]) {
  final before = calculateNetSalary(currentGross, dependents, year);
  final after = calculateNetSalary(newGross, dependents, year);
  final deltaGross = roundMoney(newGross - currentGross);
  final deltaNet = roundMoney(after.net - before.net);
  final deltaIr = roundMoney(after.ir - before.ir);
  final marginalNetRate =
      deltaGross == 0 ? 0.0 : roundMoney(deltaNet / deltaGross);

  return RaiseSimulation(
    year: year,
    dependents: dependents,
    before: before,
    after: after,
    deltaGross: deltaGross,
    deltaNet: deltaNet,
    deltaIr: deltaIr,
    marginalNetRate: marginalNetRate,
  );
}
